Add possibility to use fields instead of column for unique constraint and indexes (#8345)

Lustmored opened the pull request:
This PR implements feature requested in #8345. It handles all existing drivers (probably upcoming Annotation will need a little change if it gets merged).

I based my work on last comment by @beberlei - it translates fields and columns to indexes only in `SchemaTool::getSchemaFromMetadata`. I haven't found the need to translate them beforehand, but maybe I've missed something.

I have adapted metadata tests to cover all cases when reading entities and added simple test for translating fields to columns.

In my implementation one can use both `fields` and `columns` at the same time and they will be merged.

Lustmored committed:
Add possibility to use fields instead of column for unique constraint and indexes (#8345)

Lustmored committed:
Document changes in annotation reference

Lustmored committed:
phpcs

beberlei commented on lib/Doctrine/ORM/Tools/SchemaTool.php:
```diff
@@ -138,6 +138,37 @@ private function processingNotRequired($class, array $processedClasses)
+        $columns = [];
+
+        if (isset($indexData['columns'])) {
+            $columns = $indexData['columns'];
```
I think its not a good idea to merge them, because in case of many the order of columns is significant for how a B-Tree index is created in the database. 

The only solution where we could support mixing them would be to have a `@IndexField` and `@IndexColumn` annotation and in xml nested children `<index-field name="foo"><index-column name="bar" />`. This is a bit overkill for this problem, so I would rather that we throw an exception if you define both instead of merging them.

What do you think?

Lustmored replied:
I have adapted code to enforce exactly one of fields/columns and extended tests to cover all cases I could come up with.

beberlei commented on lib/Doctrine/ORM/Tools/SchemaTool.php:
```diff
@@ -138,6 +138,37 @@ private function processingNotRequired($class, array $processedClasses)
+        $columns = [];
+
+        if (isset($indexData['columns'])) {
+            $columns = $indexData['columns'];
```
I think its not a good idea to merge them, because in case of many the order of columns is significant for how a B-Tree index is created in the database. 

The only solution where we could support mixing them would be to have a `@IndexField` and `@IndexColumn` annotation and in xml nested children `<index-field name="foo"><index-column name="bar" />`. This is a bit overkill for this problem, so I would rather that we throw an exception if you define both instead of merging them.

What do you think?

Lustmored replied:
I have adapted code to enforce exactly one of fields/columns and extended tests to cover all cases I could come up with.

beberlei reviewed (changes requested):
I am unsure if merging is a good behavior, see inline comment.

Lustmored commented:
> I am unsure if merging is a good behavior, see inline comment.

Thank you for the review. I see the problem and agree with your point. My initial idea was to allow unmapped columns to be added to indexes, but in such corner case user would probably create unmapped index also, so this point seems invalid.

There are a few things here that will need refactor. I will do them as noted below and come back with updated PR unless there are objections:

- when to throw exception: I believe exceptions should be thrown as soon as wrong behavior is found. In this case it'd be in drivers, where it's very simple to determine that both attributes are set. I would also add exception in `SchemaTool` to future-proof the feature (if some driver in the future would allow setting both of them it should still throw)
- In XML schema as far as I know it's not possible to mark "one of 2 elements are required", therefore I would change schema to make both fields optional
- Another question is whether drivers should throw when neither `columns` nor `fields` are specified? I believe they should, as I see no use case for annotating indexes without specified columns.

Lustmored committed:
Ensure exactly one of fields/columns is set for index/uniqueConstraint

Lustmored committed:
Adapt docs to fields/columns changes

Lustmored committed:
phpcs

Lustmored committed:
Implement fields in Attribute driver and fix mapping classes constructors.

Lustmored committed:
Coding Standard

Lustmored commented:
I have rebased onto 2.9 and added implementation to new attribute driver. I am not a fan of code duplication in metadata drivers when checking fields/columns configuration, but wasn't sure how to mitigate it. Any suggestions are welcome :)
